src/main/scala/is/hail/asm4s/FunctionBuilder.scala

```diff
@@ -108,53 +157,59 @@ class FunctionBuilder[F >: Null](parameterTypeInfo: Array[MaybeGenericTypeInfo[_
 
-  def newLocal[T](implicit tti: TypeInfo[T]): LocalRef[T] =
-    newLocal()
+  def newLocal[T](name: String = null)(implicit tti: TypeInfo[T]): LocalRef[T] = apply.newLocal[T](name)
 
-  def newLocal[T](name: String = null)(implicit tti: TypeInfo[T]): LocalRef[T] =
-    new LocalRef[T](allocLocal[T](name))
+  def getArg[T](i: Int)(implicit tti: TypeInfo[T]): LocalRef[T] = apply.getArg[T](i)
 
-  def getArg[T](i: Int)(implicit tti: TypeInfo[T]): LocalRef[T] = {
-    assert(i >= 0)
-    assert(i < layout.length)
-    new LocalRef[T](argIndex(i))
-  }
+  def emit(c: Code[_]) = apply.emit(c)
 
-  val l = new mutable.ArrayBuffer[AbstractInsnNode]()
-  def emit(c: Code[_]) {
-    c.emit(l)
-  }
+  def emit(insn: AbstractInsnNode) = apply.emit(insn)
 
-  def emit(insn: AbstractInsnNode) {
-    l += insn
+  def newMethod(argsInfo: Array[MaybeGenericTypeInfo[_]], returnInfo: MaybeGenericTypeInfo[_]): MethodBuilder = {
+    val mb = new MethodBuilder(this, s"method${ methods.size }", argsInfo, returnInfo)
+    methods.append(mb)
+    mb
   }
 
+  def newMethod[A: TypeInfo, R: TypeInfo]: MethodBuilder =
+    newMethod(Array[MaybeGenericTypeInfo[_]](GenericTypeInfo[A]), GenericTypeInfo[R])
+
+  def newMethod[A: TypeInfo, B: TypeInfo, R: TypeInfo]: MethodBuilder =
+    newMethod(Array[MaybeGenericTypeInfo[_]](GenericTypeInfo[A], GenericTypeInfo[B]), GenericTypeInfo[R])
+
+  def newMethod[A: TypeInfo, B: TypeInfo, C: TypeInfo, R: TypeInfo]: MethodBuilder =
+    newMethod(Array[MaybeGenericTypeInfo[_]](GenericTypeInfo[A], GenericTypeInfo[B], GenericTypeInfo[C]), GenericTypeInfo[R])
+
+  def newMethod[A: TypeInfo, B: TypeInfo, C: TypeInfo, D: TypeInfo, R: TypeInfo]: MethodBuilder =
+    newMethod(Array[MaybeGenericTypeInfo[_]](GenericTypeInfo[A], GenericTypeInfo[B], GenericTypeInfo[C], GenericTypeInfo[D]), GenericTypeInfo[R])
+
+  def newMethod[A: TypeInfo, B: TypeInfo, C: TypeInfo, D: TypeInfo, E: TypeInfo, R: TypeInfo]: MethodBuilder =
+    newMethod(Array[MaybeGenericTypeInfo[_]](GenericTypeInfo[A], GenericTypeInfo[B], GenericTypeInfo[C], GenericTypeInfo[D], GenericTypeInfo[E]), GenericTypeInfo[R])
+
   def classAsBytes(print: Option[PrintWriter] = None): Array[Byte] = {
-    mn.instructions.add(start)
-    val dupes = l.groupBy(x => x).map(_._2.toArray).filter(_.length > 1).toArray
-    assert(dupes.isEmpty, s"some instructions were repeated in the instruction list: ${dupes: Seq[Any]}")
-    l.foreach(mn.instructions.add _)
-    mn.instructions.add(new InsnNode(returnTypeInfo.base.returnOp))
-    mn.instructions.add(end)
+    def writeInsnsToMethod(m: MethodBuilder) {
+      m.mn.instructions.add(m.start)
+      val dupes = m.l.groupBy(x => x).map(_._2.toArray).filter(_.length > 1).toArray
+      assert(dupes.isEmpty, s"some instructions were repeated in the instruction list: ${ dupes: Seq[Any] }")
+      m.l.foreach(m.mn.instructions.add _)
+      m.mn.instructions.add(new InsnNode(returnTypeInfo.base.returnOp))
+      m.mn.instructions.add(m.end)
+    }
```

Comment:
I'd put this on the MethodBuilder itself.

Reply:
I had this there originally, but it shouldn't ever get used outside of this writer and so I didn't especially want to expose it...

...Thinking about this further, what if MethodBuilder isn't attached to a FunctionBuilder? The only thing this would affect is 1. you'd have to add the method to the list of methods on a FunctionBuilder, and the invoke would probably be defined on the function builder instead. But maybe the utility of having little floating method blobs is outweighed by how confusing this would get...

Reply:
It does seem that right now it should only be called here.

I think of this as `close` or `finish` because it actually writes the instructions to the method node. Given the fact that nearly every line references the MethodBuilder, it seems much more natural to be on the MethodBuilder class.

It seems OK to call `finish` whenever the method is finished. It doesn't _need_ to happen immediately before we convert the class to bytes. It just needs to happen sometime before.

I tend to agree the MB and FB should be separate (and FB should be ClassB), but there is an inherent two-way dependency because of `invoke`. I much prefer `invoke` on the method builder, so I think I like MB having a reference to an FB.

Reply:
ok, sounds good.